kimchi/src/prover_index.rs

```diff
@@ -21,7 +21,7 @@ use std::sync::Arc;
 //~spec:startcode
 pub struct ProverIndex<G: CommitmentCurve> {
     /// constraints system polynomials
-    #[serde(bound = "ConstraintSystem<ScalarField<G>>: Serialize + DeserializeOwned")]
+    #[serde(skip)]
     pub cs: ConstraintSystem<ScalarField<G>>,
```

Comment:
oh, so that should definitely not be `skip` (I feel like I might be contradicting myself, did I say it should be `skip`?)

Instead the `precomputations` should be `skip`. But this introduces a new problem: now we need to know when something is set or not set after deserializing. I think there's a few options:

* either we set precomputations to be an `Option<Arc<...>>` and `unwrap()` it wherever we use it. So when you deserialize it is set to `None` automatically, and you have to manually call a function to set it (maybe two functions to offer the two different features: own it, or share it)
* or we set precomputations to be an `OnceCell<Arc<...>>` and if there's nothing in there when we access it we create it. But we also give the option to set the Arc. I think this is the cleaner path, but I'm not sure if I explained this very clearly :)

Reply:
something like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f4c82cfe98a03f59c5d7aa42039cf333

Reply:
i did something slightly different. i created a method `get_or_set_domain_constant_evaluation`. what do you think?